Ridge and Lasso Regression uygulamamız için modelimizi kaydedeceğiz

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd

In [3]:
# Veriyi Yükleyip EDA

In [4]:
df = pd.read_csv("Advertising.csv")

In [5]:
df.head()

Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  200 non-null    int64  
 1   TV          200 non-null    float64
 2   Radio       200 non-null    float64
 3   Newspaper   200 non-null    float64
 4   Sales       200 non-null    float64
dtypes: float64(4), int64(1)
memory usage: 7.9 KB


In [7]:
df.isnull().sum()

Unnamed: 0    0
TV            0
Radio         0
Newspaper     0
Sales         0
dtype: int64

In [8]:
# Varsa gereksiz index sütununu çıkarıyoruz

In [9]:
if 'Unnamed: 0' in df.columns:
    df = df.drop('Unnamed: 0', axis=1)

In [10]:
X = df[['TV', 'Radio', 'Newspaper']]

In [11]:
y = df['Sales']

In [12]:
# Train/Test Ayrımı

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [15]:
# Ölçeklendirme

In [16]:
from sklearn.preprocessing import StandardScaler

In [17]:
scaler = StandardScaler()

In [18]:
X_train_scaled = scaler.fit_transform(X_train)

In [19]:
X_test_scaled = scaler.transform(X_test)

In [20]:
# Modelleri Eğiteceğiz

In [21]:
from sklearn.linear_model import Ridge, Lasso

In [22]:
ridge_model = Ridge(alpha=1.0) # Alpha: Ceza terimi (Ne kadar yüksekse o kadar çok baskılar)

In [23]:
ridge_model.fit(X_train_scaled, y_train)

Ridge()

In [24]:
ridge_model.score(X_train_scaled, y_train)

0.8956679557030848

In [25]:
lasso_model = Lasso(alpha=0.1) # Lasso biraz daha agresiftir

In [26]:
lasso_model.fit(X_train_scaled, y_train)

Lasso(alpha=0.1)

In [27]:
lasso_model.score(X_train_scaled, y_train)

0.8948665120975132

In [28]:
# Değerlendirme

In [29]:
print("--- Ridge Regression ---")

--- Ridge Regression ---


In [30]:
y_pred_ridge = ridge_model.predict(X_test_scaled)

In [31]:
from sklearn.metrics import mean_squared_error, r2_score

In [32]:
print(f"R2 Score: {r2_score(y_test, y_pred_ridge):.4f}")

R2 Score: 0.8988


In [33]:
print(f"Katsayılar: {ridge_model.coef_}")

Katsayılar: [3.74169256 2.77331224 0.0633801 ]


In [34]:
print("\n--- Lasso Regression ---")


--- Lasso Regression ---


In [35]:
y_pred_lasso = lasso_model.predict(X_test_scaled)

In [36]:
print(f"R2 Score: {r2_score(y_test, y_pred_lasso):.4f}")

R2 Score: 0.8983


In [37]:
print(f"Katsayılar: {lasso_model.coef_}")

Katsayılar: [3.66920574 2.71914667 0.        ]


In [38]:
# Save

In [39]:
import joblib

In [40]:
joblib.dump(ridge_model, "ridge_model.pkl")

['ridge_model.pkl']

In [41]:
joblib.dump(lasso_model, "lasso_model.pkl")

['lasso_model.pkl']

In [42]:
joblib.dump(scaler, "ads_scaler.pkl")

['ads_scaler.pkl']

Ridge Model Score: 0.8957

Lasso Model Score: 0.8949

Ridge R2 Score: 0.8988

Lasso R2 Score: 0.8983